src/node.py

```diff
@@ -33,10 +49,44 @@ def update_term(self, term):
         if term > self.term:
             self.term = term
             self.change_type("F")
+            self.votes_received = set()
             self.voted_for = None
+            self.election_timout = self.calculate_election_timeout()
 
     def receive(self,sender,message):
-        pass
+        self.update_term(message.term)
+        assert type(message) == AppendEntries or \
+               type(message) == RequestVote or \
+               type(message) == AppendEntriesResponse or \
+               type(message) == RequestVoteResponse
+        if type(message) == AppendEntries:
+            if self.node_type == "F":
+                self.broker.add_timer(self.node_id, self.election_timeout)
+            elif self.node_type == "C" or self.node_type == "L":
+                self.change_type("F")
+        elif type(message) == RequestVote:
+            if message.term < self.term or self.voted_for is not None:
+                self.broker.send_to(self.node_id, sender, RequstVoteResponse(self.term, False))
+            else:
+                self.broker.send_to(self.node_id, sender, RequstVoteResponse(self.term, True))
```

Comment:
Shouldn't we check that candidate's log is at least as up-to-date as receiver's log?

Reply:
Yes, we should :)

Thanks!